hotsos/core/ycheck/common.py

```diff
@@ -0,0 +1,260 @@
+from hotsos.core.ycheck.engine.properties.search import CommonTimestampMatcher
+
+
+class SearchRegistryKeyConflict(Exception):
+    def __init__(self, key, all_keys):
+        self.key = key
+        self.all_keys = all_keys
+
+    def __str__(self):
+        return ("'{}' key already exists in search registry. Available keys "
+                "are:\n      - {}".
+                format(self.key, '\n      - '.join(self.all_keys)))
+
+
+class SearchRegistryKeyNotFound(Exception):
+    def __init__(self, key, all_keys):
+        self.key = key
+        self.all_keys = all_keys
+
+    def __str__(self):
+        return ("'{}' not found in search registry. Available keys are:"
+                "\n      - {}".
+                format(self.key, '\n      - '.join(self.all_keys)))
+
+
+class GlobalSearchRegistry(UserDict):
+    """
+    Maintains a set of properties e.g. dot paths to events or scenarios in yaml
+    tree - that have been registered as having a search property, a global
+    FileSearcher object and the results from running searches. This information
+    is used to load searches from a set of events, run them and save their
+    results for later retrieval. Search results are tagged with the names
+    stored here.
+    """
+
+    def __init__(self):
+        self._global_searcher = None
+        self._global_searcher_results = None
+        super().__init__()
+
+    def __setitem__(self, key, item):
+        if key in self:
+            raise SearchRegistryKeyConflict(key, list(self.data))
+
+        log.debug("adding key=%s to search registry", key)
+        super().__setitem__(key, item)
+
+    def __getitem__(self, key):
+        try:
+            return super().__getitem__(key)
+        except KeyError:
+            raise SearchRegistryKeyNotFound(key, list(self.data)) from KeyError
+
+    def get_global_searcher_results(self):
+        if self._global_searcher is None:
+            raise Exception("registry global searcher is None")
+
+        if self._global_searcher_results is not None:
+            log.debug("using cached global searcher results")
+            return self._global_searcher_results
+
+        log.debug("fetching global searcher results")
+        self._global_searcher_results = self._global_searcher.run()
+        return self._global_searcher_results
+
+    def get_global_searcher(self, allow_create=False):
+        if self._global_searcher:
+            log.debug("using existing global searcher (%s)",
+                      self._global_searcher)
+            return self._global_searcher
+
+        if not allow_create:
+            raise Exception("global searcher is not set but is "
+                            "expected to be.")
+
+        constraint = SearchConstraintSearchSince(
+                                         ts_matcher_cls=CommonTimestampMatcher)
+        searcher = FileSearcher(constraint=constraint)
+        self._global_searcher = searcher
```

Comment:
this might be for legacy reasons that no longer apply, I will try moving it back to __init__

Reply:
ok i looked into this and until we fully get rid of the global var we cant do this. i plan to do this in a follow up patch because it will mean too much change for a single patch. i have cleaned this one up a lot now though.